### PR TITLE
Pull language server URL from meta tag

### PR DESCRIPTION
**Pull request by kntsoriano**

### CODE REVIEW

**Review by iHiD — APPROVED** [review]
I'm "approving" the Gemfile + haml 🙂  Someone else can make sense of the crazy React 😉

File: app/javascript/components/student/editor/ExercismMonacoEditor.tsx [hunk 1]
```diff
@@ -44,18 +44,17 @@ export function ExercismMonacoEditor({
...
-  ) => {
-    editor.addAction({
-      id: 'runTests',
-      label: 'Run tests',
-      keybindings: [monacoEditor.KeyCode.F2],
-      run: onRunTests,
-    })
+  const startLanguageClient = useCallback(() => {
+    const languageServerUrl = document.querySelector<HTMLMetaElement>(
+      'meta[name=language-server-url]'
```
Comment (SleeplessByte) [inline app/javascript/components/student/editor/ExercismMonacoEditor.tsx]:
```suggestion
      'meta[name="language-server-url"]'
```

Both are correct, but quotes make it harder to break when changed.

File: app/javascript/components/student/editor/ExercismMonacoEditor.tsx [hunk 2]
```diff
@@ -44,18 +44,17 @@ export function ExercismMonacoEditor({
...
+    )?.content
```
Comment (SleeplessByte) [inline app/javascript/components/student/editor/ExercismMonacoEditor.tsx:50]: It always _must_ be present. You could actually throw when it's not! (I would). I know you log an error a little bit later, but I would throw.

File: app/javascript/components/student/editor/ExercismMonacoEditor.tsx [hunk 3]
```diff
@@ -88,6 +87,27 @@ export function ExercismMonacoEditor({
         connection.onClose(() => disposable.dispose())
       },
     })
+  }, [document, listen])
+
+  useEffect(() => {
+    window.addEventListener('load', startLanguageClient)
```
Comment (SleeplessByte) [inline app/javascript/components/student/editor/ExercismMonacoEditor.tsx]: Won't work with Turbolinks. What's the goal here? It seems that you'll always want to start it if the component is active? If so, do you need an event? I'm not asking these questions so you will change it, but rather trying to figure out your thinking so we can figure out what's best together 🔢

File: app/javascript/components/student/editor/ExercismMonacoEditor.tsx [hunk 4]
```diff
@@ -44,18 +44,17 @@ export function ExercismMonacoEditor({
...
+    )?.content
 
-    MonacoServices.install(editor)
-    const url = normalizeUrl(`${process.env.LANGUAGE_SERVER_HOST}/${language}/${uuidv4()}`)
+    if (!languageServerUrl) {
+      console.error('Language server URL not found')
+      return
+    }
+
+    const url = normalizeUrl(`${languageServerUrl}/${language}/${uuidv4()}`)
```
Comment (SleeplessByte) [inline app/javascript/components/student/editor/ExercismMonacoEditor.tsx]: I would probably move the entire "find and construct the url (this line, all the way up to the start of the callback)" into a `useMemo` instead, and make this thing a `useEffect` with the `url` as dependency. This also allows you to remove the `window` effect.

File: app/javascript/components/student/editor/ExercismMonacoEditor.tsx [hunk 5]
```diff
@@ -88,6 +87,27 @@ export function ExercismMonacoEditor({
         connection.onClose(() => disposable.dispose())
```
Comment (SleeplessByte) [inline app/javascript/components/student/editor/ExercismMonacoEditor.tsx:90]: But it's never closed? Turning this into an effect solves this.

File: app/javascript/components/student/editor/ExercismMonacoEditor.tsx [hunk 6]
```diff
@@ -88,6 +87,27 @@ export function ExercismMonacoEditor({
...
       },
     })
+  }, [document, listen])
+
+  useEffect(() => {
+    window.addEventListener('load', startLanguageClient)
+
+    return () => {
+      window.removeEventListener('load', startLanguageClient)
```
Comment (SleeplessByte) [inline app/javascript/components/student/editor/ExercismMonacoEditor.tsx]: Also problematic with this approach: it doesn't "disconnect" on "unload". Probably is what you wanted, but that's not what's happening.

File: app/javascript/components/student/editor/ExercismMonacoEditor.tsx (hunk 2 above)
Comment (kntsoriano) [reply app/javascript/components/student/editor/ExercismMonacoEditor.tsx:50]: Thanks! I am throwing the error now.

File: app/javascript/components/student/editor/ExercismMonacoEditor.tsx (hunk 3 above)
Comment (kntsoriano) [reply app/javascript/components/student/editor/ExercismMonacoEditor.tsx]: Your approach is a lot better and it works. I did this as I was concerned that the meta tag wouldn't be loaded by the time the React component is loaded, but didn't need to worry about that.

File: app/javascript/components/student/editor/ExercismMonacoEditor.tsx (hunk 5 above)
Comment (kntsoriano) [reply app/javascript/components/student/editor/ExercismMonacoEditor.tsx:90]: Ive now put it in an effect and the websocket properly.

**Pull request merged**